File: src/core/parser.ts
// src/core/parser.ts
//
// Capture syntax parser: tokenizer and full capture input parser.
//
// Grammar:
//   input     = (token | text)* ["--" text]
//   token     = tag | source | urgent | important | note
//   tag       = "#" word
//   source    = "@" word
//   urgent    = "!!"
//   important = "!!!"
//   note      = "--note" quoted_string | "--note" word
//   escape    = "\#" (literal # in text)
//   separator = "--" (everything after " -- " is plain text)

export interface ParsedCapture {
  text: string;
  tags: string[];
  source: string | null;
  urgent: boolean;
  important: boolean;
  notes: string[];
}

/**
 * tokenize splits a string by whitespace, respecting double-quoted strings.
 * Quotes are delimiters and are not included in the resulting tokens.
 * An unclosed quote causes the rest of the string to be one token.
 * Returns [] for empty or whitespace-only input.
 */
export function tokenize(s: string): string[] {
  s = s.trim();
  if (s === '') return [];

  const tokens: string[] = [];
  let current = '';
  let inQuote = false;

  for (let i = 0; i < s.length; i++) {
    const ch = s[i];
    if (ch === '"') {
      inQuote = !inQuote;
    } else if (ch === ' ' && !inQuote) {
      if (current.length > 0) {
        tokens.push(current);
        current = '';
      }
    } else {
      current += ch;
    }
  }

  if (current.length > 0) {
    tokens.push(current);
  }

  return tokens;
}

/**
 * extractNote finds and removes --note "value" or --note value from the raw string.
 * Returns [remaining, note]. Can be called multiple times for multiple notes.
 */
function extractNote(s: string): [string, string | null] {
  const flag = '--note';
  const idx = s.indexOf(flag);
  if (idx === -1) return [s, null];

  const before = s.slice(0, idx);
  const after = s.slice(idx + flag.length).trimStart();

  if (after === '') {
    // --note at end with no value
    return [before.trimEnd(), null];
  }

  let note: string;
  let rest: string;

  if (after[0] === '"') {
    // Quoted note value — find closing quote
    const closeIdx = after.indexOf('"', 1);
    if (closeIdx === -1) {
      // No closing quote; take the rest (without leading quote)
      note = after.slice(1);
      rest = '';
    } else {
      note = after.slice(1, closeIdx);
      rest = after.slice(closeIdx + 1).trimStart();
    }
  } else {
    // Unquoted — take the next whitespace-delimited token
    const spaceIdx = after.indexOf(' ');
    if (spaceIdx === -1) {
      note = after;
      rest = '';
    } else {
      note = after.slice(0, spaceIdx);
      rest = after.slice(spaceIdx + 1);
    }
  }

  const remaining = (before + ' ' + rest).trim();
  return [remaining, note];
}

/**
 * parseCapture parses a capture input string into its structured components.
 *
 * Processing order (mirrors Go implementation):
 * 1. Extract all --note occurrences from raw string
 * 2. Split on " -- " separator (or leading "-- " at start)
 * 3. Tokenize the pre-separator portion
 * 4. For each token: normalize \! → !, then classify
 * 5. Append plain suffix (post-separator text) to text words
 * 6. Join and trim text words
 */
export function parseCapture(input: string): ParsedCapture {
  const notes: string[] = [];
  const tags: string[] = [];
  let source: string | null = null;
  let urgent = false;
  let important = false;

  // Step 1: Extract all --note occurrences from the raw string
  let raw = input;
  while (raw.includes('--note')) {
    const [remaining, note] = extractNote(raw);
    raw = remaining;
    if (note !== null) {
      notes.push(note);
    } else {
      // --note with no value — stop trying to extract more
      break;
    }
  }

  // Step 2: Split on " -- " separator or handle leading "-- "
  let preSep = raw;
  let plainSuffix = '';

  if (raw.startsWith('-- ')) {
    // Leading separator: everything after "-- " is plain text
    preSep = '';
    plainSuffix = raw.slice(3);
  } else if (raw === '--') {
    // Just "--" alone
    preSep = '';
    plainSuffix = '';
  } else {
    const sepIdx = raw.indexOf(' -- ');
    if (sepIdx !== -1) {
      preSep = raw.slice(0, sepIdx);
      plainSuffix = raw.slice(sepIdx + 4);
    } else if (raw.endsWith(' --')) {
      preSep = raw.slice(0, raw.length - 3);
      plainSuffix = '';
    }
  }

  // Step 3: Tokenize the pre-separator portion
  const tokens = tokenize(preSep);

  // Step 4: Process each token
  const textWords: string[] = [];

  for (const tok of tokens) {
    // Normalize shell-escaped exclamation marks: \! → !
    const normTok = tok.replace(/\\!/g, '!');

    if (normTok === '!!!') {
      important = true;
    } else if (normTok === '!!') {
      urgent = true;
    } else if (tok.startsWith('#')) {
      // Tag token
      const tagName = tok.slice(1);
      if (tagName !== '') {
        tags.push(tagName);
      }
      // Bare "#" with no word → ignored (not added to text)
    } else if (tok.startsWith('\\#')) {
      // Escaped hash → literal # in text
      textWords.push('#' + tok.slice(2));
    } else if (tok.startsWith('@')) {
      // Source token
      const srcName = tok.slice(1);
      if (srcName !== '') {
        source = srcName; // last @source wins
      }
      // Bare "@" with no word → ignored
    } else {
      textWords.push(tok);
    }
  }

  // Step 5: Append plain suffix text
  if (plainSuffix !== '') {
    textWords.push(plainSuffix);
  }

  // Step 6: Join and trim
  const text = textWords.join(' ').trim();

  return { text, tags, source, urgent, important, notes };
}

File: crates/core/src/parser.rs
//! Capture-syntax parser — port of v1's [`src/core/parser.ts`](../../src/core/parser.ts).
//!
//! Grammar:
//!
//! ```text
//! input     = (token | text)* [" -- " text | "-- " text | text " --"]
//! token     = tag | source | urgent | important | note
//! tag       = "#" word
//! source    = "@" word
//! urgent    = "^^"
//! important = "^^^"
//! note      = "--note" quoted | "--note" word
//! escape    = "\#word" -> literal "#word" in body text
//! ```
//!
//! Processing order (mirrors v1 byte-for-byte so the parsers can be diffed):
//!
//! 1. Extract all `--note` occurrences from the raw string (multiple supported).
//! 2. Split on `" -- "` separator (or leading `"-- "` / trailing `" --"`).
//! 3. Tokenize the pre-separator portion (whitespace-split, quote-aware).
//! 4. Classify each token: `^^^` / `^^` / `#tag` / `\#text` / `@source` / text.
//! 5. Append the post-separator plain suffix.
//! 6. Join + trim text.

use crate::model::ParsedCapture;

#[derive(Debug, thiserror::Error)]
pub enum ParseError {
    #[error("empty capture")]
    Empty,
}

/// Parse a single capture line into structured form.
///
/// # Errors
///
/// Returns [`ParseError::Empty`] if the input has no body text after stripping
/// tokens, separators, and notes.
pub fn parse(input: &str) -> Result<ParsedCapture, ParseError> {
    let mut notes: Vec<String> = Vec::new();
    let mut tags: Vec<String> = Vec::new();
    let mut source: Option<String> = None;
    let mut urgent = false;
    let mut important = false;

    // Step 1: extract --note occurrences iteratively
    let mut raw = input.to_owned();
    while raw.contains("--note") {
        let (remaining, note) = extract_note(&raw);
        raw = remaining;
        match note {
            Some(n) => notes.push(n),
            None => break,
        }
    }

    // Step 2: split on separator
    let (pre_sep, plain_suffix) = split_separator(&raw);

    // Step 3: tokenize
    let tokens = tokenize(&pre_sep);

    // Step 4: classify
    let mut text_words: Vec<String> = Vec::new();
    for tok in tokens {
        if tok == "^^^" {
            important = true;
        } else if tok == "^^" {
            urgent = true;
        } else if let Some(rest) = tok.strip_prefix("\\#") {
            text_words.push(format!("#{rest}"));
        } else if let Some(name) = tok.strip_prefix('#') {
            if !name.is_empty() {
                tags.push(name.to_owned());
            }
        } else if let Some(name) = tok.strip_prefix('@') {
            if !name.is_empty() {
                source = Some(name.to_owned());
            }
        } else {
            text_words.push(tok);
        }
    }

    // Step 5: append plain suffix
    if !plain_suffix.is_empty() {
        text_words.push(plain_suffix);
    }

    // Step 6: join + trim
    let text = text_words.join(" ").trim().to_owned();

    if text.is_empty() && tags.is_empty() && notes.is_empty() && source.is_none() {
        return Err(ParseError::Empty);
    }

    Ok(ParsedCapture {
        text,
        tags,
        source,
        urgent,
        important,
        notes,
    })
}

/// Whitespace-split, quote-aware tokenizer. Quotes are delimiters, not part
/// of the resulting tokens. Unclosed quote → rest is one token.
pub(crate) fn tokenize(s: &str) -> Vec<String> {
    let s = s.trim();
    if s.is_empty() {
        return Vec::new();
    }

    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut in_quote = false;

    for ch in s.chars() {
        if ch == '"' {
            in_quote = !in_quote;
        } else if ch == ' ' && !in_quote {
            if !current.is_empty() {
                tokens.push(std::mem::take(&mut current));
            }
        } else {
            current.push(ch);
        }
    }

    if !current.is_empty() {
        tokens.push(current);
    }

    tokens
}

/// Find and remove the first `--note <value>` occurrence. Returns
/// `(remaining_string, extracted_note)`. If `--note` exists but has no value
/// after it, returns `(remaining, None)` so the caller can break the loop.
fn extract_note(s: &str) -> (String, Option<String>) {
    const FLAG: &str = "--note";
    let Some(idx) = s.find(FLAG) else {
        return (s.to_owned(), None);
    };

    let before = &s[..idx];
    let after = s[idx + FLAG.len()..].trim_start();

    if after.is_empty() {
        return (before.trim_end().to_owned(), None);
    }

    let (note, rest) = if let Some(after_quote) = after.strip_prefix('"') {
        match after_quote.find('"') {
            Some(close) => (
                after_quote[..close].to_owned(),
                after_quote[close + 1..].trim_start().to_owned(),
            ),
            None => (after_quote.to_owned(), String::new()),
        }
    } else {
        match after.find(' ') {
            Some(space) => (after[..space].to_owned(), after[space + 1..].to_owned()),
            None => (after.to_owned(), String::new()),
        }
    };

    let remaining = format!("{before} {rest}").trim().to_owned();
    (remaining, Some(note))
}

/// Returns `(pre_separator_string, plain_suffix_string)`.
fn split_separator(raw: &str) -> (String, String) {
    if let Some(rest) = raw.strip_prefix("-- ") {
        return (String::new(), rest.to_owned());
    }
    if raw == "--" {
        return (String::new(), String::new());
    }
    if let Some(idx) = raw.find(" -- ") {
        return (raw[..idx].to_owned(), raw[idx + 4..].to_owned());
    }
    if let Some(stripped) = raw.strip_suffix(" --") {
        return (stripped.to_owned(), String::new());
    }
    (raw.to_owned(), String::new())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parsed(text: &str, tags: &[&str]) -> ParsedCapture {
        ParsedCapture {
            text: text.to_owned(),
            tags: tags.iter().map(|t| (*t).to_owned()).collect(),
            source: None,
            urgent: false,
            important: false,
            notes: Vec::new(),
        }
    }

    #[test]
    fn parses_simple_text() {
        let p = parse("buy groceries").unwrap();
        assert_eq!(p, parsed("buy groceries", &[]));
    }

    #[test]
    fn parses_tag() {
        let p = parse("buy groceries #errands").unwrap();
        assert_eq!(p, parsed("buy groceries", &["errands"]));
    }

    #[test]
    fn parses_multiple_tags() {
        let p = parse("fix nether portal #minecraft #deep-focus").unwrap();
        assert_eq!(p, parsed("fix nether portal", &["minecraft", "deep-focus"]));
    }

    #[test]
    fn parses_urgent() {
        let p = parse("call dentist ^^").unwrap();
        assert!(p.urgent);
        assert!(!p.important);
        assert_eq!(p.text, "call dentist");
    }

    #[test]
    fn parses_important() {
        let p = parse("call dentist ^^^").unwrap();
        assert!(p.important);
        assert!(!p.urgent);
        assert_eq!(p.text, "call dentist");
    }

    #[test]
    fn parses_source() {
        let p = parse("test @cli").unwrap();
        assert_eq!(p.source.as_deref(), Some("cli"));
        assert_eq!(p.text, "test");
    }

    #[test]
    fn last_source_wins() {
        let p = parse("test @first @second").unwrap();
        assert_eq!(p.source.as_deref(), Some("second"));
    }

    #[test]
    fn parses_note_quoted() {
        let p = parse("call dentist --note \"555-1234\" #health").unwrap();
        assert_eq!(p.notes, vec!["555-1234".to_owned()]);
        assert_eq!(p.text, "call dentist");
        assert_eq!(p.tags, vec!["health".to_owned()]);
    }

    #[test]
    fn parses_note_unquoted() {
        let p = parse("call dentist --note urgent").unwrap();
        assert_eq!(p.notes, vec!["urgent".to_owned()]);
    }

    #[test]
    fn parses_multiple_notes() {
        let p = parse("x --note one --note two").unwrap();
        assert_eq!(p.notes, vec!["one".to_owned(), "two".to_owned()]);
    }

    #[test]
    fn escaped_hash_becomes_literal() {
        let p = parse("check issue \\#42 on github #work").unwrap();
        assert_eq!(p.text, "check issue #42 on github");
        assert_eq!(p.tags, vec!["work".to_owned()]);
    }

    #[test]
    fn separator_passes_text_through() {
        let p = parse("-- dump the drives").unwrap();
        assert_eq!(p.text, "dump the drives");
        assert!(p.tags.is_empty());
    }

    #[test]
    fn separator_mid_string() {
        let p = parse("text -- #not-a-tag #real-tag").unwrap();
        assert_eq!(p.text, "text #not-a-tag #real-tag");
        assert!(p.tags.is_empty());
    }

    #[test]
    fn bare_hash_is_ignored() {
        let p = parse("text # more").unwrap();
        assert_eq!(p.text, "text more");
        assert!(p.tags.is_empty());
    }

    #[test]
    fn bare_at_is_ignored() {
        let p = parse("text @ more").unwrap();
        assert_eq!(p.text, "text more");
        assert!(p.source.is_none());
    }

    #[test]
    fn empty_input_errors() {
        assert!(matches!(parse(""), Err(ParseError::Empty)));
        assert!(matches!(parse("   "), Err(ParseError::Empty)));
    }

    #[test]
    fn whitespace_around_text_trimmed() {
        let p = parse("  hello world  ").unwrap();
        assert_eq!(p.text, "hello world");
    }

    #[test]
    fn quoted_string_in_text() {
        // Quotes are tokenizer delimiters: the quoted phrase becomes a single
        // token, then text tokens rejoin with single spaces. The quotes
        // themselves are stripped.
        let p = parse("say \"hello world\" loudly").unwrap();
        assert_eq!(p.text, "say hello world loudly");
    }

    #[test]
    fn capture_with_only_tag_no_body_is_valid() {
        let p = parse("#standup").unwrap();
        assert_eq!(p.text, "");
        assert_eq!(p.tags, vec!["standup".to_owned()]);
    }

    #[test]
    fn complex_real_world_capture() {
        let p = parse(
            "fix the nether portal #minecraft #deep-focus @cli ^^ --note \"obsidian needed\"",
        )
        .unwrap();
        assert_eq!(p.text, "fix the nether portal");
        assert_eq!(
            p.tags,
            vec!["minecraft".to_owned(), "deep-focus".to_owned()]
        );
        assert_eq!(p.source.as_deref(), Some("cli"));
        assert!(p.urgent);
        assert!(!p.important);
        assert_eq!(p.notes, vec!["obsidian needed".to_owned()]);
    }
}

#[cfg(test)]
mod proptests {
    use super::*;
    use proptest::prelude::*;

    proptest! {
        // Plain ascii text without any token-like punctuation should round-trip
        // identically: no tags, no source, no flags, text == trimmed input.
        #[test]
        fn plain_text_roundtrips(s in "[a-z]{1,30}( [a-z]{1,30}){0,5}") {
            let p = parse(&s).unwrap();
            prop_assert_eq!(p.text, s.trim());
            prop_assert!(p.tags.is_empty());
            prop_assert!(p.source.is_none());
            prop_assert!(!p.urgent);
            prop_assert!(!p.important);
            prop_assert!(p.notes.is_empty());
        }

        // Any number of well-formed tags appended to text are all captured.
        #[test]
        fn tags_are_captured(
            text in "[a-z]{1,20}( [a-z]{1,20}){0,3}",
            tags in proptest::collection::vec("[a-z][a-z0-9-]{0,15}", 1..6)
        ) {
            let mut input = text.clone();
            for t in &tags {
                input.push_str(" #");
                input.push_str(t);
            }
            let p = parse(&input).unwrap();
            prop_assert_eq!(p.text, text);
            prop_assert_eq!(p.tags, tags);
        }

        // Parsing never panics on arbitrary unicode input.
        #[test]
        fn parser_never_panics(s in any::<String>()) {
            let _ = parse(&s);
        }
    }
}
